### PR TITLE
Added conversion from  <spoiler>  tag to  <details class="spoiler">

### PR DESCRIPTION
**Pull request by eliduvid**

### CODE REVIEW

File: src/main/java/org/moera/node/text/HtmlProcessor.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package org.moera.node.text;
+
+import org.intellij.lang.annotations.Language;
```
Comment (smelamud) [inline src/main/java/org/moera/node/text/HtmlProcessor.java:3]: Which package this annotation comes from? Why is it needed?

File: src/main/java/org/moera/node/text/HtmlProcessor.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.nodes.Node;
+import org.jsoup.select.NodeVisitor;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class HtmlProcessor {
```
Comment (smelamud) [inline src/main/java/org/moera/node/text/HtmlProcessor.java:13]: Please, add an empty line after opening { and before closing } of every class.

File: src/main/java/org/moera/node/text/TextConverter.java [hunk 3]
```diff
@@ -18,7 +18,7 @@ private String toHtml(SourceFormat format, String source) {
             case PLAIN_TEXT:
                 return PlainTextConverter.toHtml(source);
             case HTML:
-                return source;
+                return HtmlProcessor.process(source);
             case MARKDOWN:
                 return markdownConverter.toHtml(source);
```
Comment (smelamud) [inline src/main/java/org/moera/node/text/TextConverter.java:23]: Why not to add the same postprocessing for Markdown? `<spoiler>` tag would be useful here too.

File: src/test/java/org/moera/node/text/HtmlProcessorTest.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        String expected = "I will not tell you that killer is " +
+                "<details class=\"spoiler\">\n" +
+                "    <summary>big shock</summary>gardener\n" +
+                "</details>, it's just unethical.";
+
+        String actual = HtmlProcessor.process(source);
+
+        HtmlMatcher.assertEqHtml(expected, actual);
+    }
+}
```
Comment (smelamud) [inline src/test/java/org/moera/node/text/HtmlProcessorTest.java:40]: Newline.

**Review by smelamud — CHANGES_REQUESTED** [review]

File: src/main/java/org/moera/node/text/TextConverter.java (hunk 3 above)
Comment (eliduvid) [reply src/main/java/org/moera/node/text/TextConverter.java:23]: It wouldn't. All html tags in markdown are interpreted as block and spoiler should be inline. I'll just add `||spoiler||` to markdown with one more PR.

File: src/main/java/org/moera/node/text/HtmlProcessor.java (hunk 1 above)
Comment (eliduvid) [reply src/main/java/org/moera/node/text/HtmlProcessor.java:3]: I will remove it. It makes sense only for IntelliJ users

File: src/main/java/org/moera/node/text/TextConverter.java (hunk 3 above)
Comment (smelamud) [reply src/main/java/org/moera/node/text/TextConverter.java:23]: No. There is a special list of block tags (see `MarkdownConverted.htmlBlockTags()`), all other tags are inline.

File: src/main/java/org/moera/node/text/HtmlProcessor.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+        @Override
+        public void head(Node node, int depth) {
+            if (node instanceof Element && "spoiler".equals(node.nodeName())) {
+                spoilers.add((Element) node);
+            }
+        }
+
+        @Override
+        public void tail(Node node, int depth) {
```
Comment (smelamud) [inline src/main/java/org/moera/node/text/HtmlProcessor.java:41]: This method override is not needed.

File: src/main/java/org/moera/node/text/HtmlProcessor.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            Element details = new Element("details").addClass("spoiler");
+            if (spoiler.childrenSize() == 0 || !"summary".equals(spoiler.child(0).tagName())) {
+                details.appendChild(new Element("summary").text("spoiler!"));
+            }
+            spoiler.replaceWith(details.insertChildren(details.childNodeSize(), spoiler.childNodes()));
+        }
+        return document.body().html();
+    }
+
+    private static class SpoilerCollector implements NodeVisitor {
```
Comment (smelamud) [inline src/main/java/org/moera/node/text/HtmlProcessor.java:30]: Empty line here.

File: src/main/java/org/moera/node/text/HtmlProcessor.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        public void head(Node node, int depth) {
+            if (node instanceof Element && "spoiler".equals(node.nodeName())) {
+                spoilers.add((Element) node);
+            }
+        }
+
+        @Override
+        public void tail(Node node, int depth) {
+
+        }
```
Comment (smelamud) [inline src/main/java/org/moera/node/text/HtmlProcessor.java:43]: And here.

**Review by smelamud — CHANGES_REQUESTED** [review]

File: src/main/java/org/moera/node/text/TextConverter.java [hunk 8]
```diff
@@ -16,9 +16,9 @@
     private String toHtml(SourceFormat format, String source) {
         switch (format) {
             case PLAIN_TEXT:
-                return PlainTextConverter.toHtml(source);
+                return HtmlProcessor.process(PlainTextConverter.toHtml(source));
```
Comment (smelamud) [inline src/main/java/org/moera/node/text/TextConverter.java]: Plain text should not be touched.

Comment (eliduvid) [reply src/main/java/org/moera/node/text/TextConverter.java]: Already fixed, that was an error

File: src/main/java/org/moera/node/text/HtmlProcessor.java (hunk 5 above)
Comment (eliduvid) [reply src/main/java/org/moera/node/text/HtmlProcessor.java:41]: What should I do instead? It's an abstract method.

Comment (smelamud) [reply src/main/java/org/moera/node/text/HtmlProcessor.java:42]: Ah... OK

**Review by smelamud — APPROVED** [review]

File: src/test/java/org/moera/node/test/HtmlMatcher.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    private static Element normalizeHtml(String html) {
+        Element root = new Element("div");
+        return root.insertChildren(0, Parser.parseFragment(html, root, ""));
+    }
+
+    @Override
+    public boolean matches(Object o) {
+        if (!(o instanceof String)) return false;
```
Comment (smelamud) [inline src/test/java/org/moera/node/test/HtmlMatcher.java:26]: Please, format as multiline with braces.

File: src/main/java/org/moera/node/text/HtmlProcessor.java (hunk 5 above)
Comment (smelamud) [reply src/main/java/org/moera/node/text/HtmlProcessor.java:42]: Just remove the empty line.

```
public void tail(...) {
}
```

**Review by smelamud — CHANGES_REQUESTED** [review]

File: src/test/java/org/moera/node/test/HtmlMatcher.java (hunk 9 above)
Comment (eliduvid) [reply src/test/java/org/moera/node/test/HtmlMatcher.java:26]: OK

**Pull request merged**